XMonad/Actions/Menu/Core.hs

```diff
@@ -0,0 +1,116 @@
+{-|
+
+Module      : XMonad.Actions.Menu.Core
+Description : The core functionality that runs menus.
+Copyright   : (C) David Janssen, 2016
+License     : BSD3
+Maintainer  : David Janssen <janssen.dhj@gmail.com>
+Stability   : unstable
+Portability : unportable
+
+This module contains the code that is responsible for listening to the keyboard
+and running mappings between key-strokes and actions.
+
+I am thinking just now (at the time of making this 'publication-ready') that
+'doing-things-in-the-context-of-the-keyboard', which is defined by withKB and
+the various KeyStroke based actions is probably a Monad (since they're supposed
+to be 'computational-contexts', if I understand correctly.) I could probably
+refactor this into something nicer with that.
```

Comment:
If there are certain `X` actions that *should only* be run in the context of `withKB`, then it could make sense to create a monad which is just a newtype wrapper around `X`:
```
{-# LANGUAGE GeneralizedNewtypeDeriving #-}

newtype KB a = KB (X a)
  deriving (Functor, Applicative, Monad, blahditty blah blah all the instances that X has)

waitKey :: KB KeyStroke
waitKey = KB $ do  ... etc.

withKB :: KB a -> X (Maybe a)
```
That way the types make it clear that you have to run these things in the context of a grabbed keyboard (indeed, they make it *necessary*).